File: src/main/java/ordilov/lissn/track/infrastructure/TrackStoreImpl.java
package ordilov.lissn.track.infrastructure;

import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import ordilov.lissn.track.domain.Track;
import ordilov.lissn.track.domain.TrackStore;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class TrackStoreImpl implements TrackStore {

  private final TrackRepository trackRepository;

  @Override
  public List<Track> store(List<Track> tracks) {
    return trackRepository.saveAll(tracks);
  }

  @Override
  public Track store(Track track) {
    Optional<Track> trackOptional = trackRepository.findByResourceId(track.getResourceId());
    if(trackOptional.isEmpty()){
      return trackRepository.save(track);
    }
    return trackOptional.get();
  }
}

File: src/main/java/ordilov/lissn/track/infrastructure/TrackStoreImpl.kt
package ordilov.lissn.track.infrastructure;

import ordilov.lissn.track.domain.Track
import ordilov.lissn.track.domain.TrackStore
import org.springframework.stereotype.Component

@Component
class TrackStoreImpl(
    val trackRepository: TrackRepository
) : TrackStore {

    override fun store(tracks: List<Track>): MutableList<Track> {
        return trackRepository.saveAll(tracks)
    }

    override fun store(track: Track): Track {
        return trackRepository.findByResourceId(track.resourceId)
            .orElse(trackRepository.save(track))
    }
}